src/data/finset/noncomm_prod.lean

```diff
@@ -217,6 +217,42 @@ by simp [noncomm_prod, insert_val_of_not_mem ha, multiset.noncomm_prod_cons']
     (λ x hx y hy, by rw [mem_singleton.mp hx, mem_singleton.mp hy]) = f a :=
 by simp [noncomm_prod, multiset.singleton_eq_cons]
 
+@[to_additive]
+lemma noncomm_prod_map {α : Type*} {β : Type*} [monoid β]
+  (s : finset α) (f : α → β)
+  (comm : ∀ (x : α), x ∈ s → ∀ (y : α), y ∈ s → commute (f x) (f y))
+  {M : Type*} [monoid M] (g : β →* M) :
```

Comment:
Can you generalize this in the same way that `commute.map` is generalized?

Reply:
Generalized lemmas as far as possible, but this looses dot-notation in a few cases, and in one case a lemma wouldn’t apply anymore if I don’t use `change` before to help lean a bit.

Is this how it should be?

Also, I don’t fully understand when to use `→*` and when `monoid_hom_class`? The docs say that one should always phrase lemmas in terms of the latter, but there are plenty of lemmas that don’t – is that a recent change that hasn’t been applied everywhere yet?

Reply:
The fallout is rather big. I'd suggest I go back to the version I had before, so that this PR doesn't conflate adding new lemmas with refactoring existing ones, and we leave an overhaul of existing API to use `monoid_hom_class` for another time. Would that be ok with you, @eric-wieser?

Reply:
> Also, I don’t fully understand when to use `→*` and when `monoid_hom_class`? The docs say that one should always phrase lemmas in terms of the latter, but there are plenty of lemmas that don’t – is that a recent change that hasn’t been applied everywhere yet?

Exactly, this was changed only at the end of last year and we still need to update all the hundreds of usages.

I'd be happy to make a few separate refactoring PRs if you feel that would be too much for this one. It's definitely time to tackle `map_{list,multiset,finset,finsupp}_{prod,sum}`.

Reply:
I'd like to keep this out of the critical path for my stuff, as it's rather tangential, thanks.

Is this then good to go as is?

Reply:
Ah you approved, so someone can bors it now?

Reply:
`map_{list,multiset,finset,finsupp}_prod`: #12354

Reply:
Just to check: Does this mean the generalization is now trivial, or are there upstream lemmas that still need generalizing first?

Reply:
Just tried: all the lemmas are in place now, so all you need to do is change the lemmas that you're adding in this PR.